File: src/Day07.kt

// 6392012777720
fun main() {
    val input = readInput("day07.txt")
    var result = 0L
    for (line in input) {
        val value = line.substringBefore(":").toLong()
        val operands = line.substringAfter(":").trim().split(" ").map { it.toLong() }
        if (solvable(value, operands)) {
            result += value
        }
    }
    println(result)

    // Part 2
    result = 0L
    for (line in input) {
        val value = line.substringBefore(":").toLong()
        val operands = line.substringAfter(":").trim().split(" ").map { it.toLong() }
        if (solvable2(value, operands)) {
            result += value
        }
    }
    println(result)
}

fun solvable2(result: Long, operands: List<Long>): Boolean {
    fun trySolve(acc: Long, idx: Int): Boolean {
        if (idx == operands.size) return acc == result
        return trySolve(acc + operands[idx], idx + 1)
                || trySolve(acc * operands[idx], idx + 1)
                || trySolve((acc.toString() + operands[idx]).toLong(), idx + 1)
    }

    return trySolve(operands[0], 1)
}

fun solvable(result: Long, operands: List<Long>): Boolean {
    fun trySolve(acc: Long, idx: Int): Boolean {
        if (idx == operands.size) return acc == result
        return trySolve(acc + operands[idx], idx + 1)
                || trySolve(acc * operands[idx], idx + 1)
    }

    return trySolve(operands[0], 1)
}

File: src/day07.rs
use log::log;
use crate::read_lines;

pub fn solve() {
    let input = read_lines("day07.txt");

    let mut result = 0;
    for line in input.iter() {
        let i = line.find(":").unwrap();
        let value = line[0..i].parse::<i64>().unwrap();
        let ops = line[i + 2..]
            .trim()
            .split(" ")
            .map(|v| v.parse::<i64>().unwrap())
            .collect::<Vec<_>>();
        if try_solve(value, ops[0], 1, &ops) {
            result += value;
        }
    }

    println!("{}", result);

    let mut result = 0;
    for line in input.iter() {
        let i = line.find(":").unwrap();
        let value = line[0..i].parse::<i64>().unwrap();
        let ops = line[i + 2..]
            .trim()
            .split(" ")
            .map(|v| v.parse::<i64>().unwrap())
            .collect::<Vec<_>>();
        if try_solve2(value, ops[0], 1, &ops) {
            result += value;
        }
    }

    println!("{}", result);
}

fn try_solve2(result: i64, acc: i64, idx: usize, operands: &Vec<i64>) -> bool {
    if idx == operands.len() {
        return acc == result;
    }

    let v = operands[idx];
    let log10 = ((v as f64).log10() + 0.5).round() as i64;
    let pow = 10_i64.pow(log10.try_into().unwrap());
    try_solve2(result, acc + operands[idx], idx + 1, operands)
        || try_solve2(result, acc * operands[idx], idx + 1, operands)
        || try_solve2(result, acc * pow + operands[idx], idx + 1, operands)
}


fn try_solve(result: i64, acc: i64, idx: usize, operands: &Vec<i64>) -> bool {
    if idx == operands.len() {
        return acc == result;
    }
    try_solve(result, acc + operands[idx], idx + 1, operands)
        || try_solve(result, acc * operands[idx], idx + 1, operands)
}
